java/src/main/java/com/microsoft/tunnels/websocket/WebSocketSession.java

```diff
@@ -81,13 +85,13 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
   }
 
   @Override
-  public void suspendRead()  {
-    super.suspendRead();
+  public void suspendRead() {
+    this.reading.tryAcquire();
   }
 
   @Override
   public void resumeRead()  {
-    super.resumeRead();
+    this.reading.release();
```

Comment:
What if `suspendRead` couldn't acquire the semaphore?

Reply:
If it cant acquire the semaphore that means that suspendRead was already called. The tryAcquire will always run before Acquire

Reply:
if the semaphore is available when resumeRead is called it just returns immediately